packages/mobx/src/types/observablemap.ts

```diff
@@ -306,14 +306,14 @@ export class ObservableMap<K = any, V = any>
     }
 
     /** Merge another object into this object, returns this. */
-    merge(other: ObservableMap<K, V> | IKeyValueMap<V> | any): ObservableMap<K, V> {
+    merge(other?: IObservableMapInitialValues<K, V>): ObservableMap<K, V> {
         if (isObservableMap(other)) {
             other = new Map(other)
         }
         transaction(() => {
             if (isPlainObject(other))
                 getPlainObjectKeys(other).forEach((key: any) =>
-                    this.set(key as any as K, other[key])
+                    this.set(key as any as K, (other as any)[key])
```

Comment:
This new `as any` was introduced because of the `undefined` above.

Reply:
`other as IKeyValueMap`

I wonder if it would be correct for `isPlainObject` to also return something like `value is Object`. The compiler then should know `other[key]` is fine, right?

Reply:
I've tried with no success. Consider this example:

```js
const a: Object = [];
const b: {[k:string]: any} = [1,2,3]; 
const c: Record<string, any> = [1,2,3]; 
const d: object = [1,2,3]; 
```

None of these lines give a compilation error: https://www.typescriptlang.org/play?#code/MYewdgzgLgBAhgLhgeQEYCsCmxYF4YDaAugNwBQoksqSA3gQNYLQBOAlmAOZFJxgCeAXxj4CARgA0AJgkBmUjArhoMYEgBK2ECwAmAHlYdOE+AIB8IwpJnySiyip1IQGbHivS5CskA

It's a bit surprising but it works like this due to JS nature. Running `typeof []` in JS gives `object`. And TS `object` "represents any non-primitive type": https://mariusschulz.com/blog/the-object-type-in-typescript

Why I mentioned this? Because if I make `isPlainObject` return something like `value is object` other TS issues appear:

<img width="751" alt="Screen Shot 2022-02-10 at 15 52 09" src="https://user-images.githubusercontent.com/22447849/153412192-ce65db02-a839-4a19-8f1c-3e31ef6c92e2.png">

The type of `other` became `never` because TS narrowed its type to a primitive type. And primitive type can't be an array.

To summarize. We can either leave it as is, or add `if (other === undefined) return` to avoid any, or change `as any` to `as IKeyValueMap`

Reply:
Ah, I absolutely did not realize it also works for the `else` branch, assuming the opposite of `Object` :)
I was not familiar with this `thing as Type` construct, I am asking largely for my own educational pruposes, so thank you for the elaborate answer.

> merging with undefined doesn't make much sense anyway

Not sure if it's comparable, but I find `Object.assign` allowing for `undefined`s quite handy sometimes.

> as any to as IKeyValueMap

Let's go with that, even the `any` is fine with me. `IKeyValueMap` just seemed more correct since that's what this `isPlainObject` is about...

Reply:
I do prefer braces here (perhaps in general), but I am unopinionated about eslint rule. However, if you decide to, it will probably involve more changes, so you may want to do it as part of seperate PR, to keep diff focused.